### PR TITLE
Add anonymization command

### PR DESCRIPTION
**Pull request by phijma-leukeleu**
Ik vond dit zelf de meest logische plek. Geen idee of dit een juiste aanpak is though.

Voorbeeld gebruik in 99gram: https://github.com/leukeleu/99gram/compare/main...example-implementation-anonymization.

@wadevries Dennis ziet graag een commando (`Makefile`) waarmee je een productie DB kopieert => anonimiseert => exporteert naar staging. Valt dit onder de mogelijkheden? Zou je daarmee kunnen helpen eventueel?

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -20,7 +20,9 @@ python_requires = >=3.6
 packages = find:
 install_requires =
     Django
+    Faker
     pyyaml
+    tqdm
```
Comment (phijma-leukeleu) [inline setup.cfg:25]: Nu hier. Wat zou een alternatief zijn? Geen progressie zien bij een script dat mogelijk lang duurt lijkt me ook wenselijk 🤔

File: leukeleu_django_gdpr/anonymize.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                            fieldtype_overrides[field_type],
+                        )
+                    except KeyError:
+                        raise ImproperlyConfigured(
+                            f"Field type '{field_type}' not defined "
+                            "inside FIELDTYPE_FAKER_MAPPING"
+                        )
+
+                    for obj in qs:
+                        setattr(obj, field_name, value_func())
```
Comment (wadevries) [inline leukeleu_django_gdpr/anonymize.py:68]: Misschien hier "empty" values met rust laten? Of dus checken in condition in L46?

File: leukeleu_django_gdpr/anonymize.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+        with transaction.atomic():
+            models = get_models_from_gdpr_yml()
+            for model_name, model_data in models.items():
+                print(f"Currently anonymizing: {model_name}")
+
+                Model = apps.get_model(model_name)  # noqa: N806
+
+                qs = qs_overrides.get(model_name, Model._base_manager.all())
+                qs = qs.all()  # Makes sure we are always dealing with the latest data
```
Comment (wadevries) [inline leukeleu_django_gdpr/anonymize.py:42]: Doet dit echt iets? Wat gebeurt er als je deze regel weg laat?

Comment (phijma-leukeleu) [reply leukeleu_django_gdpr/anonymize.py:42]: Deze (hypothetische?) situatie 😄 

```python
class Anonymizer(BaseAnonymizer):
    qs_overrides = {"Foo": Foo.objects.all()}

a = Anonymizer()
a.anonymize() # Alles geanonimiseerd

Foo.objects.create()

a.anonymize()  # Alles behalve Foo net aangemaakte geanonimiseerd
```

Overigens doet Django met cbv's (`queryset = ...`) hetzelfde.

Comment (jaap3) [inline leukeleu_django_gdpr/anonymize.py:42]: Zou dit ook werken?

```suggestion
                qs = qs_overrides.get(model_name, Model._base_manager).all()
```

Comment (phijma-leukeleu) [reply leukeleu_django_gdpr/anonymize.py:42]: Zeker, maar nu kon ik er mooi commentje achter zetten

Comment (jaap3) [reply leukeleu_django_gdpr/anonymize.py:42]: Die comment past er ook boven ;)

File: leukeleu_django_gdpr/anonymize.py [hunk 4]
```diff
@@ -38,8 +39,8 @@ def anonymize(self):
 
                 Model = apps.get_model(model_name)  # noqa: N806
 
-                qs = qs_overrides.get(model_name, Model._base_manager.all())
-                qs = qs.all()  # Makes sure we are always dealing with the latest data
+                # Calling .all() makes sure we are always dealing with the latest data
```
Comment (phijma-leukeleu) [inline leukeleu_django_gdpr/anonymize.py:42]: geniet even van hoe mooi die zinnen alignen

**Comment by phijma-leukeleu** [comment]
What can I do to get this approved? 🙏

File: leukeleu_django_gdpr/anonymize.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+                    try:
+                        value_func = field_overrides.get(
+                            field_path,
+                            fieldtype_overrides[field_type],
+                        )
+                    except KeyError:
+                        raise ImproperlyConfigured(
+                            f"Field type '{field_type}' not defined "
+                            "inside FIELDTYPE_FAKER_MAPPING"
```
Comment (jaap3) [inline leukeleu_django_gdpr/anonymize.py:65]: What is `FIELDTYPE_FAKER_MAPPING`?

**Review by jaap3 — COMMENTED** [review]
Could use some docs on how to use this, and how it can be customized per project.

**Comment by phijma-leukeleu** [comment]
> Could use some docs on how to use this, and how it can be customized per project.
I was asked to create a notion page:
https://www.notion.so/leukeleu/Data-anonymisation-cd1a48e3d4514108b854c4b4f5df794e?pvs=4

**Comment by phijma-leukeleu** [comment]
@dbunskoek @bheesink 

Discussed with @jaap3 and @wadevries:

There is a management command (this PR) that anonymizes the current database.

The steps to get an anonymized production database on staging are:
* `make get-db` => but with `l_<project>@...` instead of `s_<project>@...`
* run `docker compose exec python ./manage.py anonymize`
* put it on staging, manually
Documentation about these steps will appear in the knowledge base.

This seems good enough:
* the people who'll do this are probably familiar with the steps involved
* this is not a process that's repeated multiple times per week (or month probably)

After it all seems to work out as expected, we can always "automize" it more.

File: README.md [hunk 6]
```diff
@@ -123,6 +123,49 @@ include:
...
+class Anonymizer(BaseAnonymizer):
+    # To skip rows
+    extra_qs_overrides = {
+        "auth.User": User._base_manager.exclude(is_superuser=True),
+        ...
+    }
+
+    # To change what fake data should be used for a field
+    extra_field_overrides = {
+        "auth.User.first_name": fake.first_name,
```
Comment (jaap3) [inline README.md:152]: Would be nice to mention where `fake` is supposed to come from. Or maybe use an `@property` so `self.fake` can be used? (still mention that it's an instance of `Faker`).

File: README.md [hunk 7]
```diff
@@ -123,6 +123,49 @@ include:
...
+        ...
+    }
+
+    # To skip fields
+    excluded_fields = [
+        "auth.User.email",  # Usually you should not exclude emails
+        ...
+    ]
```
Comment (jaap3) [inline README.md:160]: Should we even support this? What's the use case for this? When would you mark a field as PII in gdpr.yml, but then explicitly exclude it here?

File: README.md [hunk 8]
```diff
@@ -123,6 +123,49 @@ include:
...
+By default, it will anonymize **all fields** that contain PII.
+
+To change the configuration, you can create a subclass of `BaseAnonymizer`:
+
+```python
+# some_file.py
+class Anonymizer(BaseAnonymizer):
+    # To skip rows
+    extra_qs_overrides = {
+        "auth.User": User._base_manager.exclude(is_superuser=True),
```
Comment (jaap3) [inline README.md:146]: Probably nice to mention that the default behaviour does this (and also excludes staff)

File: README.md (hunk 6 above)
Comment (jaap3) [reply README.md:152]: oh and also mention that this particular example already is used by default

File: README.md (hunk 7 above)
Comment (phijma-leukeleu) [reply README.md:160]: https://github.com/leukeleu/99gram/compare/main...example-implementation-anonymization#diff-3b705f52cc3de32e432c9cb26758704b6d855ce433c01ed2b113310422e11179R32-R40

These were actual requirements for 99gram

Comment (jaap3) [reply README.md]: Interesting, I'd say that these fields are misclassified but maybe I don't fully understand the reasoning

**Review by jaap3 — APPROVED** [review]
Looks pretty good to me, I'm not too concerned about the `excluded_fields` feature. So I'm approving this now, please wait for additional review by @wadevries before merging.

**Review by wadevries — APPROVED** [review]
LGTM!

**Pull request merged**